src/gui/qgsnewvectorlayerdialog.cpp

```diff
@@ -150,30 +149,48 @@ void QgsNewVectorLayerDialog::mFileFormatComboBox_currentIndexChanged( int index
 void QgsNewVectorLayerDialog::mTypeBox_currentIndexChanged( int index )
 {
   // FIXME: sync with providers/ogr/qgsogrprovider.cpp
+
+  // ESRI field width specification is taken from here
+  // https://webhelp.esri.com/arcgisdesktop/9.3/index.cfm?TopicName=Geoprocessing%20considerations%20for%20shapefile%20output
+
+  constexpr int ESRI_TEXT_MAX_WIDTH_INCLUSIVE = 254;
+  constexpr int ESRI_SHORT_INTEGER_MAX_WIDTH_INCLUSIVE = 4;
+  // constexpr int ESRI_LONG_INTEGER_MAX_WIDTH_INCLUSIVE = 9;
+  // constexpr int ESRI_FLOAT_MAX_WIDTH_INCLUSIVE = 13;
+  constexpr int ESRI_DOUBLE_MAX_WIDTH_INCLUSIVE = 13;
+  constexpr int ESRI_DATE_MAX_WIDTH_INCLUSIVE = 8;
+
   switch ( index )
   {
     case 0: // Text data
-      if ( mWidth->text().toInt() < 1 || mWidth->text().toInt() > 255 )
-        mWidth->setText( QStringLiteral( "80" ) );
+      mWidth->setMaximum( ESRI_TEXT_MAX_WIDTH_INCLUSIVE );
+      if ( mWidth->value() < 1 || mWidth->value() > ESRI_TEXT_MAX_WIDTH_INCLUSIVE )
+        mWidth->setValue( 80 );
       mPrecision->setEnabled( false );
-      mWidth->setValidator( new QIntValidator( 1, 255, this ) );
       break;
 
     case 1: // Whole number
-      if ( mWidth->text().toInt() < 1 || mWidth->text().toInt() > 10 )
-        mWidth->setText( QStringLiteral( "10" ) );
+      mWidth->setMaximum( ESRI_SHORT_INTEGER_MAX_WIDTH_INCLUSIVE );
+      if ( mWidth->value() < 1 || mWidth->value() > ESRI_SHORT_INTEGER_MAX_WIDTH_INCLUSIVE )
+        mWidth->setValue( ESRI_SHORT_INTEGER_MAX_WIDTH_INCLUSIVE );
```

Comment:
Is this actually required? I believe it should be handled by QSpinBox in the setMaximum call already.

Reply:
Ah, yes, i see!
Changed QSpinBox to QgsSpinBox, undone this check and used ClearValue as additional tweak